Bug fix #793; use deterministic option name; can't rely on list.sort

ericfrederich opened the pull request:
Please merge; this matches the implementation to the documentation.

Currently an option defined as...
```python
@click.option('--ambitious', '--all', '-a', is_flag=True)
```
...will target `all` instead of `ambitious`

untitaker commented:
Thanks, please:

* retarget at 6.x-maintenance branch
* Write a changelog and testcase

Also I think the fix you have in mind can have a more minimal diff. In particular all that needs to be done is to replace `-1` with `0` I think.

ericfrederich commented:
@untitaker taking 0 would take from the beginning of the list which would be one of the short options.

I looked into it and it appears `sorted()` and `list.sort()` are documented as __*stable*__ in that *it guarantees not to change the relative order of elements that compare equal*.  So because this is documented, and not just an implementation detail, it is okay to rely on.  We don't have to do comprehension like the current pull request is showing.

... but it's not as simple as taking 0 either.  I'll amend this commit with one with a smaller diff.

ericfrederich commented:
Hows that for a __*more minimal diff*__?  If not for the comment the size of the code would remain unchanged ;-)... just moved the minus sign up a line and changed a 1 to a 0 😄

untitaker commented:
That looks really good, please add a changelog + test though

On Fri, May 19, 2017 at 06:05:44AM -0700, Eric L. Frederich wrote:
> Hows that for a __*more minimal diff*__?  If not for the comment the size of the code would remain unchanged ;-)... just moved the minus sign up a line and changed a 1 to a 0 😄 
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/pallets/click/pull/794#issuecomment-302697567

ericfrederich commented:
Okay @untitaker Got the changelog and test as well.

ericfrederich commented:
@untitaker whenever this is merged I'd like to create another pull request adding more testing and documentation around the explicit naming of options.  I'd prefer to keep them as separate commits and base it on this one.

untitaker commented:
Thank you very much! Feel free to open a new PR regarding new tests (preferrably targeted at master).

ericfrederich commented:
@untitaker I created #798 and #799
